server/tserver/src/main/java/org/apache/accumulo/tserver/NativeMapCleanerUtil.java

```diff
@@ -43,14 +44,37 @@ public static Cleanable deleteNM(Object obj, Logger log, AtomicLong nmPtr) {
     });
   }
 
+  // Chose 7 cleaners because each cleaner creates a thread, so do not want too many threads. This
+  // should reduce lock contention for scans by a 7th vs a single cleaner, so that is good
+  // reduction and 7 does not seem like too many threads to add. This array is indexed using
+  // pointers addresses from native code, so there is a good chance those are memory aligned on
+  // a multiple of 4, 8, 16, etc. So if changing the array size avoid multiples of 2.
+  private static final Cleaner[] NMI_CLEANERS = new Cleaner[7];
+
+  static {
+    for (int i = 0; i < NMI_CLEANERS.length; i++) {
+      NMI_CLEANERS[i] = Cleaner.create();
+    }
+  }
+
   public static Cleanable deleteNMIterator(Object obj, AtomicLong nmiPtr) {
     requireNonNull(nmiPtr);
-    return CleanerUtil.CLEANER.register(obj, () -> {
+
+    long ptr = Math.abs(nmiPtr.get());
+    if (ptr == Long.MIN_VALUE) {
+      ptr = 0;
+    }
+    int cleanerIndex = (int) (ptr % NMI_CLEANERS.length);
```

Comment:
You can avoid the absolute value and handling of the `Long.MIN_VALUE` by just fixing Java's negative answer for modulus:

```suggestion
    int cleanerIndex = (int) (nmiPtr.get() % NMI_CLEANERS.length);
    if (cleanerIndex < 0) {
      cleanerIndex += NMI_CLEANERS.length;
    }
```